File: dart/test/bandeira_test.dart
import 'dart:convert';
import 'dart:io';

import 'package:bandeira/bandeira.dart';
import 'package:test/test.dart';

Map<String, dynamic> _loadFixtures() {
  final raw = File('../testdata/flags.json').readAsStringSync();
  final decoded = jsonDecode(raw);
  if (decoded is Map<String, dynamic>) {
    return decoded;
  }
  if (decoded is Map) {
    return decoded.map((key, value) => MapEntry(key.toString(), value));
  }
  throw StateError('fixtures must decode to a JSON object');
}

final Map<String, dynamic> _fixtures = _loadFixtures();

BandeiraClient createClientFromFixtures() {
  final client = BandeiraClient(
    const BandeiraConfig(
      url: 'http://localhost:9999',
      token: 'test-token',
    ),
  );
  client.loadFlags(_fixtures);
  return client;
}

Future<HttpServer> newTestServer(Map<String, dynamic> responseBody) async {
  final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
  server.listen((request) async {
    if (request.uri.path != '/api/v1/flags') {
      request.response.statusCode = HttpStatus.notFound;
      await request.response.close();
      return;
    }

    if (request.headers.value(HttpHeaders.authorizationHeader) !=
        'Bearer test-token') {
      request.response.statusCode = HttpStatus.unauthorized;
      request.response.write('unauthorized');
      await request.response.close();
      return;
    }

    request.response.headers.contentType = ContentType.json;
    request.response.write(jsonEncode(responseBody));
    await request.response.close();
  });
  return server;
}

void main() {
  group('BandeiraClient', () {
    group('basic toggles', () {
      test('enabled flag with no strategies returns true', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('simple-on'), isTrue);
      });

      test('disabled flag returns false', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('simple-off'), isFalse);
      });

      test('unknown flag returns false', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('nonexistent'), isFalse);
      });
    });

    group('default strategy', () {
      test('returns true', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('default-strategy'), isTrue);
      });
    });

    group('userWithId', () {
      test('matches a listed user', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'user-targeting',
            const BandeiraContext(userId: 'user-42'),
          ),
          isTrue,
        );
      });

      test('rejects an unlisted user', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'user-targeting',
            const BandeiraContext(userId: 'user-99'),
          ),
          isFalse,
        );
      });

      test('rejects when no context provided', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('user-targeting'), isFalse);
      });

      test('handles newline-separated user IDs', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'user-targeting-newlines',
            const BandeiraContext(userId: 'user-42'),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'user-targeting-newlines',
            const BandeiraContext(userId: 'user-99'),
          ),
          isFalse,
        );
      });
    });

    group('gradualRollout', () {
      test('100% rollout is always on', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'rollout-100',
            const BandeiraContext(userId: 'anyone'),
          ),
          isTrue,
        );
      });

      test('0% rollout is always off', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'rollout-0',
            const BandeiraContext(userId: 'anyone'),
          ),
          isFalse,
        );
      });

      test('no userId means no stickiness', () {
        final client = createClientFromFixtures();
        expect(client.isEnabled('rollout-50'), isFalse);
      });

      test('session stickiness uses sessionId', () {
        final client = createClientFromFixtures();
        final result = client.isEnabled(
          'rollout-session-stickiness',
          const BandeiraContext(sessionId: 'sess-123'),
        );
        expect(result, isA<bool>());
      });
    });

    group('remoteAddress', () {
      test('exact match', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'ip-allowlist',
            const BandeiraContext(remoteAddress: '10.0.0.1'),
          ),
          isTrue,
        );
      });

      test('prefix match', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'ip-allowlist',
            const BandeiraContext(remoteAddress: '192.168.1.100'),
          ),
          isTrue,
        );
      });

      test('no match', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'ip-allowlist',
            const BandeiraContext(remoteAddress: '172.16.0.1'),
          ),
          isFalse,
        );
      });

      test('legacy IPs key works', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'ip-allowlist-legacy',
            const BandeiraContext(remoteAddress: '10.0.0.1'),
          ),
          isTrue,
        );
      });
    });

    group('constraints', () {
      test('IN operator matches', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-in',
            const BandeiraContext(properties: <String, String>{'companyId': '2'}),
          ),
          isTrue,
        );
      });

      test('IN operator rejects', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-in',
            const BandeiraContext(properties: <String, String>{'companyId': '99'}),
          ),
          isFalse,
        );
      });

      test('NOT_IN operator matches', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-not-in',
            const BandeiraContext(properties: <String, String>{'plan': 'enterprise'}),
          ),
          isTrue,
        );
      });

      test('NOT_IN operator rejects', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-not-in',
            const BandeiraContext(properties: <String, String>{'plan': 'free'}),
          ),
          isFalse,
        );
      });

      test('inverted constraint', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-inverted',
            const BandeiraContext(properties: <String, String>{'plan': 'free'}),
          ),
          isFalse,
        );
        expect(
          client.isEnabled(
            'constraint-inverted',
            const BandeiraContext(
              properties: <String, String>{'plan': 'enterprise'},
            ),
          ),
          isTrue,
        );
      });

      test('case-insensitive constraint', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-case-insensitive',
            const BandeiraContext(properties: <String, String>{'country': 'brazil'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-case-insensitive',
            const BandeiraContext(properties: <String, String>{'country': 'PORTUGAL'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-case-insensitive',
            const BandeiraContext(properties: <String, String>{'country': 'spain'}),
          ),
          isFalse,
        );
      });

      test('STR_CONTAINS', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-str-contains',
            const BandeiraContext(properties: <String, String>{'email': 'user@acme.com'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-str-contains',
            const BandeiraContext(properties: <String, String>{'email': 'user@other.com'}),
          ),
          isFalse,
        );
      });

      test('STR_STARTS_WITH', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-str-starts-with',
            const BandeiraContext(properties: <String, String>{'email': 'admin@acme.com'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-str-starts-with',
            const BandeiraContext(properties: <String, String>{'email': 'user@acme.com'}),
          ),
          isFalse,
        );
      });

      test('STR_ENDS_WITH', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-str-ends-with',
            const BandeiraContext(properties: <String, String>{'email': 'user@acme.com'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-str-ends-with',
            const BandeiraContext(properties: <String, String>{'email': 'user@acme.io'}),
          ),
          isFalse,
        );
      });

      test('NUM_GTE', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-num-gte',
            const BandeiraContext(properties: <String, String>{'age': '21'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-num-gte',
            const BandeiraContext(properties: <String, String>{'age': '18'}),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-num-gte',
            const BandeiraContext(properties: <String, String>{'age': '16'}),
          ),
          isFalse,
        );
      });

      test('DATE_AFTER', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constraint-date-after',
            const BandeiraContext(
              properties: <String, String>{'signupDate': '2026-06-15T00:00:00Z'},
            ),
          ),
          isTrue,
        );
        expect(
          client.isEnabled(
            'constraint-date-after',
            const BandeiraContext(
              properties: <String, String>{'signupDate': '2025-06-15T00:00:00Z'},
            ),
          ),
          isFalse,
        );
      });
    });

    group('multi-strategy', () {
      test('VIP user matches first strategy', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'multi-strategy',
            const BandeiraContext(userId: 'vip-1'),
          ),
          isTrue,
        );
      });
    });

    group('constrained rollout', () {
      test('passes when constraint matches', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constrained-rollout',
            const BandeiraContext(
              userId: 'any-user',
              properties: <String, String>{'companyId': 'acme'},
            ),
          ),
          isTrue,
        );
      });

      test('fails when constraint does not match', () {
        final client = createClientFromFixtures();
        expect(
          client.isEnabled(
            'constrained-rollout',
            const BandeiraContext(
              userId: 'any-user',
              properties: <String, String>{'companyId': 'other'},
            ),
          ),
          isFalse,
        );
      });
    });

    group('allFlags', () {
      test('returns all flags with their enabled state', () {
        final client = createClientFromFixtures();
        final flags = client.allFlags();
        expect(flags['simple-on'], isTrue);
        expect(flags['simple-off'], isFalse);
      });
    });

    group('validation', () {
      test('throws on missing url', () {
        expect(
          () => BandeiraClient(
            const BandeiraConfig(
              url: '',
              token: 'test',
            ),
          ),
          throwsArgumentError,
        );
      });

      test('throws on missing token', () {
        expect(
          () => BandeiraClient(
            const BandeiraConfig(
              url: 'http://localhost',
              token: '',
            ),
          ),
          throwsArgumentError,
        );
      });
    });

    group('startup APIs', () {
      test('create performs initial fetch', () async {
        final server = await newTestServer(_fixtures);
        addTearDown(() async {
          await server.close(force: true);
        });

        final client = await BandeiraClient.create(
          BandeiraConfig(
            url: 'http://127.0.0.1:${server.port}',
            token: 'test-token',
            pollInterval: const Duration(hours: 1),
          ),
        );
        addTearDown(client.close);

        expect(client.isEnabled('simple-on'), isTrue);
      });

      test('start performs initial fetch', () async {
        final server = await newTestServer(_fixtures);
        addTearDown(() async {
          await server.close(force: true);
        });

        final client = BandeiraClient(
          BandeiraConfig(
            url: 'http://127.0.0.1:${server.port}',
            token: 'test-token',
            pollInterval: const Duration(hours: 1),
          ),
        );
        addTearDown(client.close);

        await client.start();

        expect(client.isEnabled('simple-on'), isTrue);
      });

      test('create fails fast when initial request fails', () async {
        final server = await HttpServer.bind(InternetAddress.loopbackIPv4, 0);
        server.listen((request) async {
          request.response.statusCode = HttpStatus.unauthorized;
          request.response.write('unauthorized');
          await request.response.close();
        });
        addTearDown(() async {
          await server.close(force: true);
        });

        final future = BandeiraClient.create(
          BandeiraConfig(
            url: 'http://127.0.0.1:${server.port}',
            token: 'test-token',
          ),
        );
        await expectLater(future, throwsA(isA<StateError>()));
      });
    });
  });
}

File: go/bandeira_test.go
package bandeira

import (
	"encoding/json"
	"net/http"
	"net/http/httptest"
	"testing"
	"time"
)

func newTestServer(flags []flag) *httptest.Server {
	return httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		if r.Header.Get("Authorization") != "Bearer test-token" {
			http.Error(w, "unauthorized", http.StatusUnauthorized)
			return
		}
		json.NewEncoder(w).Encode(apiResponse{Flags: flags})
	}))
}

func TestIsEnabled_BasicToggle(t *testing.T) {
	srv := newTestServer([]flag{
		{Name: "on-flag", Enabled: true},
		{Name: "off-flag", Enabled: false},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("on-flag") {
		t.Error("expected on-flag to be enabled")
	}
	if c.IsEnabled("off-flag") {
		t.Error("expected off-flag to be disabled")
	}
	if c.IsEnabled("unknown-flag") {
		t.Error("expected unknown flag to be disabled")
	}
}

func TestIsEnabled_DefaultStrategy(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "with-default",
			Enabled: true,
			Strategies: []strategy{
				{Name: "default"},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("with-default") {
		t.Error("expected flag with default strategy to be enabled")
	}
}

func TestIsEnabled_UserWithId(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "user-flag",
			Enabled: true,
			Strategies: []strategy{
				{Name: "userWithId", Parameters: map[string]any{"userIds": "1,2,42"}},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("user-flag", Context{UserID: "42"}) {
		t.Error("expected user 42 to match")
	}
	if c.IsEnabled("user-flag", Context{UserID: "99"}) {
		t.Error("expected user 99 to not match")
	}
	if c.IsEnabled("user-flag") {
		t.Error("expected no context to not match userWithId")
	}
}

func TestIsEnabled_GradualRollout(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "rollout-100",
			Enabled: true,
			Strategies: []strategy{
				{Name: "gradualRollout", Parameters: map[string]any{"rollout": float64(100)}},
			},
		},
		{
			Name:    "rollout-0",
			Enabled: true,
			Strategies: []strategy{
				{Name: "gradualRollout", Parameters: map[string]any{"rollout": float64(0)}},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("rollout-100", Context{UserID: "anyone"}) {
		t.Error("expected 100% rollout to be enabled")
	}
	if c.IsEnabled("rollout-0", Context{UserID: "anyone"}) {
		t.Error("expected 0% rollout to be disabled")
	}
}

func TestIsEnabled_Constraint_IN(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "constrained",
			Enabled: true,
			Strategies: []strategy{
				{
					Name: "default",
					Constraints: []constraint{
						{ContextName: "companyId", Operator: "IN", Values: []string{"1", "2", "3"}},
					},
				},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("constrained", Context{Properties: map[string]string{"companyId": "2"}}) {
		t.Error("expected companyId 2 to match IN constraint")
	}
	if c.IsEnabled("constrained", Context{Properties: map[string]string{"companyId": "99"}}) {
		t.Error("expected companyId 99 to not match")
	}
}

func TestIsEnabled_Constraint_Inverted(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "inverted",
			Enabled: true,
			Strategies: []strategy{
				{
					Name: "default",
					Constraints: []constraint{
						{ContextName: "plan", Operator: "IN", Values: []string{"free"}, Inverted: true},
					},
				},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	// "free" is in the list → inverted → false
	if c.IsEnabled("inverted", Context{Properties: map[string]string{"plan": "free"}}) {
		t.Error("expected free plan to be excluded by inverted constraint")
	}
	// "enterprise" is NOT in the list → inverted → true
	if !c.IsEnabled("inverted", Context{Properties: map[string]string{"plan": "enterprise"}}) {
		t.Error("expected enterprise plan to pass inverted constraint")
	}
}

func TestIsEnabled_UserWithId_Newlines(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "user-newline",
			Enabled: true,
			Strategies: []strategy{
				{Name: "userWithId", Parameters: map[string]any{"userIds": "1\n2\n42"}},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("user-newline", Context{UserID: "42"}) {
		t.Error("expected user 42 to match with newline-separated ids")
	}
	if c.IsEnabled("user-newline", Context{UserID: "99"}) {
		t.Error("expected user 99 to not match")
	}
}

func TestIsEnabled_Constraint_CaseInsensitive(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "ci-flag",
			Enabled: true,
			Strategies: []strategy{
				{
					Name: "default",
					Constraints: []constraint{
						{ContextName: "country", Operator: "IN", Values: []string{"Brazil", "Portugal"}, CaseInsensitive: true},
					},
				},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("ci-flag", Context{Properties: map[string]string{"country": "brazil"}}) {
		t.Error("expected case-insensitive match for 'brazil'")
	}
	if !c.IsEnabled("ci-flag", Context{Properties: map[string]string{"country": "PORTUGAL"}}) {
		t.Error("expected case-insensitive match for 'PORTUGAL'")
	}
	if c.IsEnabled("ci-flag", Context{Properties: map[string]string{"country": "spain"}}) {
		t.Error("expected 'spain' to not match")
	}
}

func TestIsEnabled_Constraint_STR_CONTAINS(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "contains-flag",
			Enabled: true,
			Strategies: []strategy{
				{
					Name: "default",
					Constraints: []constraint{
						{ContextName: "email", Operator: "STR_CONTAINS", Values: []string{"@acme.com"}},
					},
				},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("contains-flag", Context{Properties: map[string]string{"email": "user@acme.com"}}) {
		t.Error("expected 'user@acme.com' to contain '@acme.com'")
	}
	if c.IsEnabled("contains-flag", Context{Properties: map[string]string{"email": "user@other.com"}}) {
		t.Error("expected 'user@other.com' to not contain '@acme.com'")
	}
}

func TestIsEnabled_RemoteAddress(t *testing.T) {
	srv := newTestServer([]flag{
		{
			Name:    "ip-flag",
			Enabled: true,
			Strategies: []strategy{
				{Name: "remoteAddress", Parameters: map[string]any{"ips": "10.0.0.1\n192.168.1."}},
			},
		},
		{
			Name:    "ip-flag-legacy",
			Enabled: true,
			Strategies: []strategy{
				{Name: "remoteAddress", Parameters: map[string]any{"IPs": "10.0.0.1,192.168.1."}},
			},
		},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	if !c.IsEnabled("ip-flag", Context{RemoteAddress: "10.0.0.1"}) {
		t.Error("expected 10.0.0.1 to match")
	}
	if !c.IsEnabled("ip-flag", Context{RemoteAddress: "192.168.1.100"}) {
		t.Error("expected 192.168.1.100 to match prefix")
	}
	if c.IsEnabled("ip-flag", Context{RemoteAddress: "172.16.0.1"}) {
		t.Error("expected 172.16.0.1 to not match")
	}

	if !c.IsEnabled("ip-flag-legacy", Context{RemoteAddress: "10.0.0.1"}) {
		t.Error("expected legacy IPs key to work")
	}
}

func TestAllFlags(t *testing.T) {
	srv := newTestServer([]flag{
		{Name: "a", Enabled: true},
		{Name: "b", Enabled: false},
	})
	defer srv.Close()

	c, err := New(Config{URL: srv.URL, Token: "test-token", PollInterval: time.Hour})
	if err != nil {
		t.Fatal(err)
	}
	defer c.Close()

	flags := c.AllFlags()
	if !flags["a"] {
		t.Error("expected a to be true")
	}
	if flags["b"] {
		t.Error("expected b to be false")
	}
}

func TestNew_InvalidToken(t *testing.T) {
	srv := newTestServer(nil)
	defer srv.Close()

	_, err := New(Config{URL: srv.URL, Token: "wrong-token", PollInterval: time.Hour})
	if err == nil {
		t.Error("expected error with invalid token")
	}
}

func TestNew_MissingURL(t *testing.T) {
	_, err := New(Config{Token: "test"})
	if err == nil {
		t.Error("expected error with missing URL")
	}
}

func TestNew_MissingToken(t *testing.T) {
	_, err := New(Config{URL: "http://localhost"})
	if err == nil {
		t.Error("expected error with missing token")
	}
}
